### PR TITLE
[MIG] stock_vertical_lift: Migration to 14.0

### PR DESCRIPTION
**Pull request by dzungtran89**
Dependencies:
- stock_location_tray: https://github.com/OCA/stock-logistics-warehouse/pull/1079
- web_ir_actions_act_view_reload: https://github.com/OCA/web/pull/1816
- web_notify: https://github.com/OCA/web/pull/1824

### CODE REVIEW

File: stock_vertical_lift/models/vertical_lift_shuttle.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        event occurred to inform the user on their screen.
+
+        The method is private only to prevent xml/rpc calls to
+        interact with the screen.
+        """
+        # XXX do we want to do something special in the notification?
+        self._operation_for_mode()._send_notification_refresh()
+
+
+class VerticalLiftShuttleManualBarcode(models.TransientModel):
```
Comment (rousseldenis) [inline stock_vertical_lift/models/vertical_lift_shuttle.py]: Could you maybe split one model per file ?

Comment (mmequignon) [reply stock_vertical_lift/models/vertical_lift_shuttle.py]: ping @dzungtran89

Comment (hailangvn) [reply stock_vertical_lift/models/vertical_lift_shuttle.py]: Good day @rousseldenis, @mmequignon, the model has been split to https://github.com/OCA/stock-logistics-warehouse/blob/8ba7d977a5f513f934c6d23ce47f59a304972a5c/stock_vertical_lift/wizards/vertical_lift_shuttle_manual_barcode.py. This PR has been rebased and updated with latest test requirement. Thanks for your comments.

Comment (hailangvn) [reply stock_vertical_lift/models/vertical_lift_shuttle.py]: I am checking the travis...

**Comment by hailangvn** [comment]
Good day @rousseldenis, @mmequignon, the travis and runbot are green now.

File: stock_vertical_lift/models/vertical_lift_operation_base.py [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+
+def split_other_move_lines(move, move_lines):
+    """Substract `move_lines` from `move.move_line_ids`, put the result
+    in a new move and returns it.
+    """
+    other_move_lines = move.move_line_ids - move_lines
+    if other_move_lines or move.state == "partially_available":
+        qty_to_split = move.product_uom_qty - sum(move_lines.mapped("product_uom_qty"))
+        backorder_move_vals = move._split(qty_to_split)
+        backorder_move = move.create(backorder_move_vals)
```
Comment (sebalix) [inline stock_vertical_lift/models/vertical_lift_operation_base.py:27]: You have to confirm the new move:
```suggestion
        backorder_move = move.create(backorder_move_vals)
        backorder_move._action_confirm(merge=False)
```
See https://github.com/OCA/wms/blob/b0ceac26303ef8278d641fb050399d3c965ecc6c/shopfloor/models/stock_move.py#L41-L43

**Comment by hailangvn** [comment]
Good day @sebalix, thanks for your help. The PR has been updated.

**Comment by simahawk** [comment]
@hailangvn can you check the build?

**Comment by hailangvn** [comment]
> @hailangvn can you check the build?

@simahawk, thanks for your approval. The PR has been rebased to latest 14.0 and updated to remove test dependency.

**Review by mmequignon — APPROVED** [review]

File: stock_vertical_lift/models/vertical_lift_operation_base.py (hunk 2 above)
Comment (simahawk) [reply stock_vertical_lift/models/vertical_lift_operation_base.py:27]: @hailangvn what about this?  Also, are we sure we always get values? We might have an empty dict AFAIR. Can you check?

File: stock_vertical_lift/models/vertical_lift_command.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    command = fields.Char(required=True)
+    answer = fields.Char()
+    error = fields.Char()
+    shuttle_id = fields.Many2one("vertical.lift.shuttle", required=True)
+
+    def record_answer(self, answer):
+        name = self._get_key(answer)
+        record = self.search([("name", "=", name)], limit=1)
+        if not record:
+            if not getattr(threading.currentThread(), "testing", False):
```
Comment (simahawk) [inline stock_vertical_lift/models/vertical_lift_command.py]: do we really need this?

Comment (hailangvn) [reply stock_vertical_lift/models/vertical_lift_command.py]: Thanks @simahawk for your review. I will work on this.
> @hailangvn what about this? Also, are we sure we always get values? We might have an empty dict AFAIR. Can you check?

**Comment by hailangvn** [comment]
Good day @simahawk, this is the investigation.
> @hailangvn what about this? Also, are we sure we always get values? We might have an empty dict AFAIR. Can you check?
* https://github.com/odoo/odoo/blob/14.0/addons/stock/models/stock_move.py#L1599 shows that empty response only happens when (a) split qty is 0, (b) split qty >= stock quantity
(a) tested with split qty = 0: no operation is created for vertical lift shuttle so cannot run to the code.
(b.1) testing with split qty = stock quantity = 30: 3 operations. Scanning and saving each operation had no issue. Last one has no other_move_lines so the _split was not run.
(b.2) testing with split qty = 35 > stock quantity = 30: 3 operations. Last one had two lines with qty 10 and 5 correspondingly. The _split result was not empty.
* To be in safe side, a check is put in place to process when backorder is not empty.

**Comment by hailangvn** [comment]
We found that tests of `stock_location_tray` were failed because of `stock_vertical_lift`. So we need to add a fix.

**Review by sebalix — APPROVED** [review]
LG, need to squash the last commits into the `[MIG] stock_vertical_lift: Migration to 14.0` one, I pre-approve

**Comment by hailangvn** [comment]
Thanks @sebalix for the comment. The PR has been updated.

**Review by simahawk — APPROVED** [review]

**Pull request merged**